src/node_sockaddr-inl.h

```diff
@@ -0,0 +1,238 @@
+      const sockaddr_in6* ipv6 =
+          reinterpret_cast<const sockaddr_in6*>(addr);
+      const uint64_t* a =
+          reinterpret_cast<const uint64_t*>(&ipv6->sin6_addr);
+      hash_combine(&hash, ipv6->sin6_port, a[0], a[1]);
+      break;
+    }
+    default:
+      UNREACHABLE();
+  }
+  return hash;
+}
+
+bool SocketAddress::Compare::operator()(
+    const sockaddr* laddr,
+    const sockaddr* raddr) const {
+  CHECK(laddr->sa_family == AF_INET || laddr->sa_family == AF_INET6);
+  return memcmp(laddr, raddr, GetLength(laddr)) == 0;
+}
+
+bool SocketAddress::is_numeric_host(const char* hostname) {
+  return is_numeric_host(hostname, AF_INET) ||
+         is_numeric_host(hostname, AF_INET6);
+}
+
+bool SocketAddress::is_numeric_host(const char* hostname, int family) {
+  std::array<uint8_t, sizeof(struct in6_addr)> dst;
+  return inet_pton(family, hostname, dst.data()) == 1;
+}
+
+sockaddr_storage* SocketAddress::ToSockAddr(
+    int32_t family,
+    const char* host,
+    uint32_t port,
+    sockaddr_storage* addr) {
+  switch (family) {
+    case AF_INET:
+      return uv_ip4_addr(
+          host,
+          port,
+          reinterpret_cast<sockaddr_in*>(addr)) == 0 ?
+              addr : nullptr;
+    case AF_INET6:
+      return uv_ip6_addr(
+          host,
+          port,
+          reinterpret_cast<sockaddr_in6*>(addr)) == 0 ?
+              addr : nullptr;
+    default:
+      UNREACHABLE();
+  }
+}
+
+int SocketAddress::GetPort(const sockaddr* addr) {
+  return ntohs(addr->sa_family == AF_INET ?
+      reinterpret_cast<const sockaddr_in*>(addr)->sin_port :
+      reinterpret_cast<const sockaddr_in6*>(addr)->sin6_port);
+}
+
+int SocketAddress::GetPort(const sockaddr_storage* addr) {
+  return GetPort(reinterpret_cast<const sockaddr*>(addr));
+}
+
+std::string SocketAddress::GetAddress(const sockaddr* addr) {
+  char host[INET6_ADDRSTRLEN];
+  const void* src = addr->sa_family == AF_INET ?
+      static_cast<const void*>(
+          &(reinterpret_cast<const sockaddr_in*>(addr)->sin_addr)) :
+      static_cast<const void*>(
+          &(reinterpret_cast<const sockaddr_in6*>(addr)->sin6_addr));
+  uv_inet_ntop(addr->sa_family, src, host, INET6_ADDRSTRLEN);
+  return std::string(host);
+}
+
+std::string SocketAddress::GetAddress(const sockaddr_storage* addr) {
+  return GetAddress(reinterpret_cast<const sockaddr*>(addr));
+}
+
+size_t SocketAddress::GetLength(const sockaddr* addr) {
+  return
+      addr->sa_family == AF_INET6 ?
+          sizeof(sockaddr_in6) :
+          sizeof(sockaddr_in);
+}
+
+size_t SocketAddress::GetLength(const sockaddr_storage* addr) {
+  return GetLength(reinterpret_cast<const sockaddr*>(addr));
+}
+
+SocketAddress::SocketAddress(const sockaddr* addr) {
+  memcpy(&address_, addr, GetLength(addr));
+}
+
+SocketAddress::SocketAddress(const SocketAddress& addr) {
+  memcpy(&address_, &addr.address_, addr.GetLength());
+}
+
+inline SocketAddress& SocketAddress::operator=(const sockaddr* addr) {
```

Comment:
```suggestion
SocketAddress& SocketAddress::operator=(const sockaddr* addr) {
```

(`ìnline` typically goes into the header)